arangod/Cluster/ClusterMethods.cpp

```diff
@@ -3301,24 +3301,36 @@ arangodb::Result matchBackupServersSlice(VPackSlice const planServers,
   match.clear();
 
   // Local copy of my servers
-  std::unordered_set<std::string> localCopy;
+  // Note that we use a `std::set` here for the following reasons:
+  //  - this function is supposed to return a canonical result, which
+  //    does not depend on the order of the input in `planServers` or
+  //    `dbServers`, therefore we use a sorted set here and a sorted
+  //    map as the result.
+  //  - Performance does not matter at all, this method is called for
+  //    hotbackup download and hotbackup restore, both methods are
+  //    using considerable amounts of time. The server lists are usually
+  //    quite small (3 or 5 or a few dozen at most).
+  // So please do not "optimize" this.
+  std::set<std::string> localCopy;
   std::copy(dbServers.begin(), dbServers.end(),
             std::inserter(localCopy, localCopy.end()));
 
-  // Skip all direct matching names in pair and remove them from localCopy
-  std::unordered_set<std::string>::iterator it;
+  // Skip all direct matching names in pair and remove them from localCopy.
+  // If later a server does not occur it means that no translation has to
+  // happen for it.
   for (auto planned : VPackObjectIterator(planServers)) {
     auto const plannedStr = planned.key.copyString();
-    if ((it = localCopy.find(plannedStr)) != localCopy.end()) {
+    auto it = localCopy.find(plannedStr);
+    if (it != localCopy.end()) {
       localCopy.erase(it);
     } else {
       match.try_emplace(plannedStr, std::string());
     }
   }
   // match all remaining
   auto it2 = localCopy.begin();
-  for (auto& m : match) {
-    m.second = *it2++;
+  for (auto& m : match) {  // alphabetical order!
```

Comment:
Is it guaranteed here that `match` and `localCopy` have the same amount of members?
If not then `it2` could run over the bounds of the `localCopy` container.
I think we should at least add an assertion here that the number of members in `localCopy` is greater or equal to the number of members in `match`.

Reply:
Initially, we check that dbServers has at least as many members as planServers, otherwise we bail out.
`localCopy` is initially a copy of dbServers.
Then we only remove from `localCopy` things which are in `planServers`,
for those servers in planServers, which are not in localCopy, we create an entry in `match`. Therefore it follows, that in the end, localCopy has at least as many members as match, as far as I see. We could add a comment to this effect and an assertion, Will do.